### PR TITLE
Enable or disable deal watching depending on communication on blockchain

### PR DESCRIPTION
**Pull request by jbern0rd**

### CODE REVIEW

File: src/main/java/com/iexec/core/chain/BlockchainConnectionHealthIndicator.java [hunk 1]
```diff
@@ -145,19 +161,24 @@ private void connectionFailed() {
...
      * <p>
-     * If never been OUT-OF-SERVICE, then:
+     * If previous failures occurred, then:
      * <ul>
-     *     <li>Reset the {@link BlockchainConnectionHealthIndicator#firstFailure} var to {@code null};</li>
-     *     <li>Log a "connection restored" message.</li>
-     * </ul>
+     * <li>Log a "connection restored" message.
+     * <li>Reset the {@link BlockchainConnectionHealthIndicator#firstFailure} var to {@code null}
+     * <li>If OUT-OF-SERVICE, publish a {@link ChainConnectedEvent} event and reset the OUT-OF-SERVICE state
```
Comment (mcornaton) [inline src/main/java/com/iexec/core/chain/BlockchainConnectionHealthIndicator.java:168]: It seems the closing `</ul>` is recommended there, based on [the JavaDoc documentation](https://www.oracle.com/fr/technical-resources/articles/java/javadoc-tool.html).
```suggestion
     * <li>If OUT-OF-SERVICE, publish a {@link ChainConnectedEvent} event and reset the OUT-OF-SERVICE state
     * </ul>
```

File: src/main/java/com/iexec/core/chain/DealWatcherService.java [hunk 2]
```diff
@@ -90,7 +94,38 @@ public DealWatcherService(ChainConfig chainConfig,
...
+    }
+
+    /**
+     * Create new deal watching subscription.
+     */
+    @Async
+    @EventListener(ChainConnectedEvent.class)
+    public void reconnect() {
+        run();
+    }
```
Comment (mcornaton) [inline src/main/java/com/iexec/core/chain/DealWatcherService.java:109]: Is there a reason why you have 2 methods doing the exact same operations? Shouldn't we have a single `connect` method that would be triggered by both events?

Comment (jbern0rd) [reply src/main/java/com/iexec/core/chain/DealWatcherService.java]: I just did not know it was possible
see 453ab8d

File: src/main/java/com/iexec/core/chain/BlockchainConnectionHealthIndicator.java (hunk 1 above)
Comment (jbern0rd) [reply src/main/java/com/iexec/core/chain/BlockchainConnectionHealthIndicator.java:168]: see 4439335

File: src/main/java/com/iexec/core/chain/BlockchainConnectionHealthIndicator.java [hunk 3]
```diff
@@ -145,19 +161,24 @@ private void connectionFailed() {
...
+        if (consecutiveFailures > 0) {
             log.info("Blockchain connection is now restored after a period of unavailability." +
-                    " [unavailabilityPeriod:{}]", pollingInterval.multipliedBy(consecutiveFailures));
+                    " [block:{}, unavailabilityPeriod:{}]",
+                    latestBlockNumber, pollingInterval.multipliedBy(consecutiveFailures));
             firstFailure = null;
+            consecutiveFailures = 0;
+            if (outOfService) {
+                applicationEventPublisher.publishEvent(new ChainConnectedEvent(this));
+                outOfService = false;
```
[the remark this answers is not in the document]
Comment (jbern0rd) [reply src/main/java/com/iexec/core/chain/BlockchainConnectionHealthIndicator.java]: see 14f7e1d

File: src/main/java/com/iexec/core/chain/BlockchainConnectionHealthIndicator.java [hunk 4]
```diff
@@ -110,28 +120,34 @@ void scheduleMonitoring() {
      */
     void checkConnection() {
         final long latestBlockNumber = web3jService.getLatestBlockNumber();
+        log.debug("blockNumber {}", latestBlockNumber);
```
[the remark this answers is not in the document]
Comment (mcornaton) [reply src/main/java/com/iexec/core/chain/BlockchainConnectionHealthIndicator.java]: And so did I! Thanks! :)

**Pull request merged**